pretyping/glob_ops.ml

```diff
@@ -484,7 +484,8 @@ let is_gvar id c = match DAst.get c with
 | GVar id' -> Id.equal id id'
 | _ -> false
 
-let rec cases_pattern_of_glob_constr na = DAst.map (function
+let rec cases_pattern_of_glob_constr na c =
+  DAst.make (match DAst.get c with
```

Comment:
Why is the `map` removed here? Note that `map` preserves attributes in the AST whereas the current pattern does not.

Reply:
I need to force evaluation and `DAst.map` keeps "thunks" unevaluated. But you are right that the loc is lost and I re-added it. Alternatively, we may have a `map_force` in `DAst` (in anticipation of more attributes being added)?

Reply:
I guess we could add `DAst.force` instead.

Reply:
You mean writing `let rec cases_pattern_of_glob_constr na c = DAst.map (function ...) (DAst.force c)`?

Reply:
For example, or even:
```
let rec cases_pattern_of_glob_constr na = DAst.map ...

let cases_pattern_of_glob_constr na c = cases_pattern_of_glob_constr na (force c)

(* or *)
let cases_pattern_of_glob_constr na c = force (cases_pattern_of_glob_constr na c)
```
etc...

Reply:
The best would likely be to stop relying on exceptions [which seems like a field of mines in the lazy setting], but I guess that is pretty hard to do.

Reply:
The problem is that I need more than evaluation of the first level of `DAst`. There might be `Thunk` in the inner levels. Forcing only the first level was actually what I was doing before but this is not enough.

Exceptions are indeed a field of mines in a lazy setting.

Reply:
> The problem is that I need more than evaluation of the first level of `DAst`.

Ah, I see, then I guess it is OK to do as you propose [a force on each recursive call]